llvm/lib/Target/DirectX/DXILPrepare.cpp

```diff
@@ -247,9 +247,10 @@ class DXILPrepareModule : public ModulePass {
 
   DXILPrepareModule() : ModulePass(ID) {}
   void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesAll();
     AU.addPreserved<ShaderFlagsAnalysisWrapper>();
     AU.addPreserved<DXILResourceMDWrapper>();
-    AU.addPreserved<DXILMetadataAnalysisWrapperPass>();
+    AU.addRequired<DXILMetadataAnalysisWrapperPass>();
```

Comment:
Likely an opportunity for me to learn something:

If we're using `setPreservesAll()` why do we need to explicitly call `addPreserved` on any other passes?

Reply:
Is it correct to say that this preserves all? I'm not sure if any analyses cache information about function attributes, so maybe it's okay there, but we do modify the IR so I doubt it's safe to really say we preserve all analyses...

Reply:
> Is it correct to say that this preserves all? I'm not sure if any analyses cache information about function attributes, so maybe it's okay there, but we do modify the IR so I doubt it's safe to really say we preserve all analyses...

I share the concern about specifying `setPreservedAll()` for `DXILPrepare` pass that actually changes the LLVM IR (as I was discussing offline with @damyanp earlier). As things stand, `DXContainerGlobals` pass that runs after `DXILPrepare` pass uses `DXILMetadataAnalysis` info (i.e., requires the pass) but running it before `DXContainerGlobals` would produce incomplete info since `DXILPrepare` deletes function attributes. Hence the unsatisfactory workaround of `setPreservedAll()` in `DXILPrepare`. 

To address this, I'd like to move attribute removal out of `DXILPrepare` into `DXContainerGlobals` pass and (correctly) delete `setPreservedAll()` from `DXILPrepare` pass. Is there a known reason for removing function attributes in `DXILPrepare` pass and not in a later pass?

Reply:
I'll plan to move the attribute removal in a follow on PR, if there are no technical reasons to address the issue raised here.

Reply:
I still don't understand why we would change this to `setPreservesAll`. If you just need to preserve `DXILMetadataAnalysis` then the existing `AU.addPreserved<DXILMetadataAnalysisWrapperPass>()` call should do that, no?

Reply:
I'm not arguing that the `addRequired<DXILMetadataAnalysisWrapperPass>()` isn't necessary, it obviously is. I'm saying we shouldn't change the specific list of preserved passes to `setPreservesAll`. This pass does not preserve all analyses, since it modifies instructions. The code before your change looks correct to me:
```c++
    AU.addPreserved<ShaderFlagsAnalysisWrapper>();
    AU.addPreserved<DXILResourceMDWrapper>();
    AU.addPreserved<DXILMetadataAnalysisWrapperPass>();
    AU.addPreserved<DXILResourceWrapperPass>();
```

We can simply add `AU.addRequired<DXILMetadataAnalysisWrapperPass>();` after those calls instead of changing those calls to `setPreservesAll`, since that is incorrect.

Reply:
Changed.